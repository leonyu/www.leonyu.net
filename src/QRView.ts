
import * as QRCode from 'qrcode';
import { debounce } from './AsyncUtils';

export default class QRView {
  qrSvg: HTMLElement;
  qrText: HTMLInputElement;

  constructor(container: HTMLElement) {
    this.qrText = document.createElement('input');
    container.appendChild(this.qrText);
    this.qrSvg = document.createElement('div');
    container.appendChild(this.qrSvg);
    this.updateInput('');
    this.qrText.addEventListener('input', debounce(() => { this.updateInput(this.qrText.value); }, 50));
  }

  updateInput(text: string): void {
    this.updateTextBox(text);
    this.updateQRCode(text);
  }

  private async updateQRCode(text: string): Promise<void> {
    const options: QRCode.QRCodeToStringOptions = { type: 'svg', color: { dark: '#222' } };

    const svgMarkup = await QRCode.toString(text || [{ data: '', mode: 'alphanumeric' }], options);
    this.qrSvg.innerHTML = svgMarkup;
  }

  private updateTextBox(text: string): void {
    if (this.qrText.value !== text) {
      this.qrText.value = text;
    }
  }
}
