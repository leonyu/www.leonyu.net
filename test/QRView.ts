import { TextEncoder } from 'node:util';
globalThis.TextEncoder ??= TextEncoder;

import { describe, afterEach, it, expect } from '@jest/globals'

import QRView from '../src/QRView'

describe('QRView', () => {
    let div: HTMLDivElement;

    it('should render correct HTML', () => {
        div = document.createElement('div');
        document.body.appendChild(div);
        const qrView = new QRView(div);
        expect(div.outerHTML).toMatchSnapshot();
        qrView.updateInput('hello world');
        expect(div.outerHTML).toMatchSnapshot();
    });

    afterEach(() => {
        div?.parentElement?.removeChild(div);
    })
});