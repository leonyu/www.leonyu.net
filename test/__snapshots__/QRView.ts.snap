// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QRView should render correct HTML 1`] = `"<div><input><div></div></div>"`;

exports[`QRView should render correct HTML 2`] = `"<div><input><div></div></div>"`;
